firedrake/assemble.py

```diff
@@ -1338,24 +1338,29 @@ def _make_maps_and_regions(self):
         test, trial = self._form.arguments()
         if self._allocation_integral_types is not None:
             return ExplicitMatrixAssembler._make_maps_and_regions_default(test, trial, self._allocation_integral_types)
-        elif any(local_kernel.indices == (None, None) for local_kernel, _ in self._all_local_kernels):
+        elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
             # Handle special cases: slate or split=False
-            assert all(local_kernel.indices == (None, None) for local_kernel, _ in self._all_local_kernels)
+            assert all(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels)
             allocation_integral_types = set(local_kernel.kinfo.integral_type
-                                            for local_kernel, _ in self._all_local_kernels)
+                                            for assembler in self._all_assemblers
+                                            for local_kernel, _ in assembler.local_kernels)
             return ExplicitMatrixAssembler._make_maps_and_regions_default(test, trial, allocation_integral_types)
         else:
             maps_and_regions = defaultdict(lambda: defaultdict(set))
-            for local_kernel, subdomain_id in self._all_local_kernels:
-                i, j = local_kernel.indices
-                mesh = self._form.ufl_domains()[local_kernel.kinfo.domain_number]
```

Comment:
The use of `self._form.ufl_domains()` is correct only in the single mesh setting. We should instead use `assembler._form.ufl_domains()`.